src/Microsoft.IdentityModel.JsonWebTokens/JsonWebToken.cs

```diff
@@ -579,14 +579,26 @@ private void ReadToken(string encodedJson)
             EncodedToken = encodedJson;
         }
 
+        /// <summary>
+        /// Serializes the <see cref="JsonWebToken"/>.
+        /// </summary>
+        /// <returns>A string containing the header and payload in JSON format.</returns>
+        public override string ToString()
+        {
+            if (!IsEncrypted)
+                return Header.ToString() + "." + Payload.ToString();
+            else
+                return Header.ToString()  + "." + EncryptedKey + "." + InitializationVector + "." + Ciphertext;
```

Comment:
Have you evaluated StringBuilder in this case? Usually when you're concatonating more than ~3 strings StringBuilder wins.

Reply:
For three string parts:
> |                          Method |       Mean |     Error |    StdDev |     Median |   Gen0 | Allocated |
|-------------------------------- |-----------:|----------:|----------:|-----------:|-------:|----------:|
| ConcatStringsUsingStringBuilder | 27.4693 ns | 1.1695 ns | 3.2793 ns | 26.5547 ns | 0.0242 |     152 B |
|          ConcatStringsUsingPlus |  0.0000 ns | 0.0000 ns | 0.0000 ns |  0.0000 ns |      - |         - |

Interesting, seems like just plussing them together still wins at 7....Benchmark.net isn't able to distinguish a difference between that and an empty method.
> |                           Method |       Mean |     Error |    StdDev |   Gen0 | Allocated |
|--------------------------------- |-----------:|----------:|----------:|-------:|----------:|
| ConcatStringsUsingStringBuilder7 | 61.3046 ns | 1.2456 ns | 1.4828 ns | 0.0459 |     288 B |
|          ConcatStringsUsingPlus7 |  0.0000 ns | 0.0000 ns | 0.0000 ns |      - |         - |

I forced in a ToString() to make sure it was actually sampling but got a similar result so seems like your code is good here.
> |                           Method |     Mean |    Error |    StdDev |   Median |   Gen0 | Allocated |
|--------------------------------- |---------:|---------:|----------:|---------:|-------:|----------:|
| ConcatStringsUsingStringBuilder7 | 97.53 ns | 3.743 ns | 10.370 ns | 93.10 ns | 0.0790 |     496 B |
|          ConcatStringsUsingPlus7 | 12.99 ns | 0.601 ns |  1.752 ns | 12.45 ns | 0.0217 |     136 B |

Reply:
I am curious as to what did you use to do your benchmark? I would like to use that tool.

About the string concatenation at hand, I found that in these type of scenarios where the string concatenation happens in a single line, the compiler adds the required optimizations such that there are no immutable strings waste here.

In this case the optimization looks like
``` csharp
string[] array = new string[7];
        array[0] = header.ToString();
        array[1] = ".";
        array[2] = EncryptedKey;
        array[3] = ".";
        array[4] = InitializationVector;
        array[5] = ".";
        array[6] = Ciphertext;
        return string.Concat(array);
```

I use this tool to retrieve the optimized code generated by the compiler, 
[https://sharplab.io/](https://sharplab.io/) I hope you find it useful.

Reply:
https://github.com/dotnet/BenchmarkDotNet

Reply:
Thanks!

Reply:
@TimHannMSFT/@victorm-hernandez  StringBuilder should be used in for loops or dynamics cases. I believe the compiler generates code similar to the string.Concat method where it calculates the total length of all strings first, allocate a memory once then copy all values. It is efficient because it does not need to re-allocate the buffer.